core/src/libraries/Oracle.sol

```diff
@@ -5,36 +5,95 @@ import {FixedPointMathLib as SoladyMath} from "solady/utils/FixedPointMathLib.so
+        uint160 sqrtMeanPriceX96;
+        // the mean liquidity over some period (OracleLibrary.consult())
+        uint160 secondsPerLiquidityX128;
+        // the number of seconds to look back when getting mean tick & mean liquidity
+        uint32 oracleLookback;
+        // the liquidity depth at currentTick (from pool.liquidity())
+        uint128 tickLiquidity;
+    }
+
     /**
      * @notice Calculates time-weighted means of tick and liquidity for a given Uniswap V3 pool
      * @param pool Address of the pool that we want to observe
+     * @param seed The indices of `pool.observations` where we start our search for the 30-minute-old (lowest 16 bits)
+     * and 60-minute-old (next 16 bits) observations. Determine these off-chain to make this method more efficient
+     * than Uniswap's binary search. If any of the highest 8 bits are set, we fallback to binary search.
+     * @return data An up-to-date `PoolData` struct containing all fields except `oracleLookback` and `tickLiquidity`
      * @return metric If the price was manipulated at any point in the past `UNISWAP_AVG_WINDOW` seconds, then at
      * some point in that period, this value will spike. It may still be high now, or (if the attacker is smart and
      * well-financed) it may have returned to nominal.
-     * @return sqrtMeanPriceX96 The sqrt(geometricMeanPrice) over the past `UNISWAP_AVG_WINDOW` seconds
-     * @return secondsPerLiquidityX128 The change in seconds per liquidity over the past `UNISWAP_AVG_WINDOW` seconds
      */
-    function consult(
-        IUniswapV3Pool pool
-    ) internal view returns (uint56 metric, uint160 sqrtMeanPriceX96, uint160 secondsPerLiquidityX128) {
-        uint32[] memory secondsAgos = new uint32[](3);
-        secondsAgos[0] = UNISWAP_AVG_WINDOW * 2;
-        secondsAgos[1] = UNISWAP_AVG_WINDOW;
-        secondsAgos[2] = 0;
-
-        (int56[] memory tickCumulatives, uint160[] memory secondsPerLiquidityCumulativeX128s) = pool.observe(
-            secondsAgos
-        );
-        secondsPerLiquidityX128 = secondsPerLiquidityCumulativeX128s[2] - secondsPerLiquidityCumulativeX128s[1];
+    function consult(IUniswapV3Pool pool, uint40 seed) internal view returns (PoolData memory data, uint56 metric) {
+        uint16 observationIndex;
+        uint16 observationCardinality;
+        (data.sqrtPriceX96, data.currentTick, observationIndex, observationCardinality, , , ) = pool.slot0();
 
         unchecked {
+            int56[] memory tickCumulatives = new int56[](3);
+            uint160[] memory secondsPerLiquidityCumulativeX128s = new uint160[](3);
+
+            if ((seed >> 32) > 0) {
+                uint32[] memory secondsAgos = new uint32[](3);
+                secondsAgos[0] = UNISWAP_AVG_WINDOW * 2;
+                secondsAgos[1] = UNISWAP_AVG_WINDOW;
+                secondsAgos[2] = 0;
+                (tickCumulatives, secondsPerLiquidityCumulativeX128s) = pool.observe(secondsAgos);
```

Comment:
If any of the highest 8 bits of `seed` are set, fallback to Uniswap's method that uses binary search (this is what we were using before).